### DISCUSSION TITLE
Representing partitioning in SQLAlchemy

### DISCUSSION THREAD

**Discussion by cmaduike** [sqlalchemy/sqlalchemy#9146] [Usage Questions] [answered]
I have a general question about partitioned tables in SqlAlchemy models, specifically postgres-based tables: What are the ramifications of not representing the underlying partitioning structure in the model classes? In other languages such as Java and the Spring framework, we can define an entity class without awareness of the partitioning structure of the table. 

Additionally, I found this works in SqlAlchemy. Here is an example of a defined partitioned table and partition in postgres:

```
CREATE TABLE test_table (
                             pk default nextval('test_table_pk_seq'::regclass) not null,
                             created_ts = Column(DateTime, server_default=text("now()")),
                             amount        int
) PARTITION BY RANGE (created_ts);

CREATE TABLE test_table_2006 PARTITION OF measurement
    FOR VALUES FROM ('2006-01-01') TO ('2007-01-01');

alter table test_table_2006 add primary key (pk);
```

I can represent this in SqlAlchemy without referencing its partitioning structure, and CRUD operations will work fine.

```
class TestTable(Base):
    __tablename__ = "test_table"

    created_ts = Column(DateTime, server_default=text("now()"))
    pk = sa.Column(sa.BIGINT, primary_key=True)
    amount = sa.Column(sa.Numeric(37, 2), nullable=True)
```

The primary key constraint in the model maps to the constraint defined on the partition in the underlying table.

I am of the mind that partitioning is such a low-level concept that the ORMs need not represent this, mainly because it complicates foreign key relationships. However I'd like to know if this should be discouraged.

**Comment by zzzeek** [comment, marked as answer]
using the partitions at the database level is encouraged :).   We have an extension called "horizontal sharding" that does a similar thing as partitioning on the client side, but it definitely is not anywhere near as robust or easy to use as PG's feature.

**Comment by cmaduike** [comment]
In that case, sticking to partitioning on the database level without modeling this in my SQLAlchemy classes should be ok and not cause any issues? I've tested this already, and it works fine, but I might be missing something.

**Reply by CaselIT** [reply]
I think the main things to be aware are:
- creating the table with sqlalchemy will not use partitioning. But if you manually create it in the db that's not a problem
- if I'm not mistaken pg will not create the partition for you, so if you try to insert something that cannot go in any partition it will raise an error. sqlalchemy will also not create them for you

**Reply by cmaduike** [reply]
Yes exactly. In my use case the partitioned table and partitions will need to be manually created in the database.

**Reply by CaselIT** [reply]
I'm not sure if a table can be made partitioned with alter.
If so you may be able to create the table with sqlalchemy then automatically alter it by registering one or more [DDL](https://docs.sqlalchemy.org/en/14/core/ddl.html#sqlalchemy.schema.DDL) to be run after table creation using the events, as in the example in the docs

**Reply by cmaduike** [reply]
Yup, existing tables cannot be converted into a partitioned table in postgres. And to your earlier point which I'd missed, I will be responsible for creating new partitions to prevent runtime errors (outside of sqlalchemy)

I'm clear on the mechanics of partitioning in postgres, but I wanted a gut check on _needing_ this represented in the SqlAlchemy models. I think I have a good idea how to proceed!

**Reply by CaselIT** [reply]
IIRC also pg mostly hides that away from you. When using a partitioned table, you can completely avoid to refer to the single partition and always use the main table in both DML and DQL statements

I'm looking at the pg dialect and it does support partition_by: https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#postgresql-table-options so you may also define in the orm `__table_args__` the partion by option and using ddl you can create the single partition tables you need

not automatic, but it removes the first point above

**Reply by cmaduike** [reply]
Yes, I've tried modeling partitioning in Sqlalchemy, but it gets too complicated when other tables have relations with this table. I'm just trying to avoid it altogether in sqlalchemy which I'm going to do based on this discussion. I've never had to model partitioning in my ORM for other languages, so I think it's safe to bypass this as well in Sqlalchemy.

**Reply by CaselIT** [reply]
I meant model only the creation, but you would still use a single orm class.
Something like this:

```py
import sqlalchemy as sa
from sqlalchemy import orm


Base = orm.declarative_base()


class Partitioned(Base):
    __tablename__ = "test_table"
    pk = sa.Column(sa.BigInteger, sa.Identity(start=1))
    created_ts = sa.Column(sa.DateTime, server_default=sa.func.now())
    ...
    __table_args__ = {"postgresql_partition_by": "RANGE (created_ts)"}
    __mapper_args__ = {"primary_key": pk}


for ddl in [
    """CREATE TABLE test_table_2006 PARTITION OF test_table
    FOR VALUES FROM ('2006-01-01') TO ('2007-01-01');
    """,
    """CREATE TABLE test_table_2007 PARTITION OF test_table
    FOR VALUES FROM ('2007-01-01') TO ('2008-01-01');
    """,
    # ...
]:
    sa.event.listen(Partitioned.__table__, "after_create", sa.DDL(ddl))


engine = sa.create_engine("postgresql://scott:tiger@/test", echo=True)
with engine.connect() as conn:
    Base.metadata.create_all(conn)
    conn.rollback()
```

this will generate the following log
```
2023-01-25 21:20:17,471 INFO sqlalchemy.engine.Engine
CREATE TABLE test_table (
        pk BIGINT GENERATED BY DEFAULT AS IDENTITY (START WITH 1),
        created_ts TIMESTAMP WITHOUT TIME ZONE DEFAULT now()
)
 PARTITION BY RANGE (created_ts)


2023-01-25 21:20:17,471 INFO sqlalchemy.engine.Engine [no key 0.00019s] {}
2023-01-25 21:20:17,482 INFO sqlalchemy.engine.Engine CREATE TABLE test_table_2006 PARTITION OF test_table
    FOR VALUES FROM ('2006-01-01') TO ('2007-01-01');

2023-01-25 21:20:17,482 INFO sqlalchemy.engine.Engine [no key 0.00012s] {}
2023-01-25 21:20:17,483 INFO sqlalchemy.engine.Engine CREATE TABLE test_table_2007 PARTITION OF test_table
    FOR VALUES FROM ('2007-01-01') TO ('2008-01-01');

2023-01-25 21:20:17,484 INFO sqlalchemy.engine.Engine [no key 0.00012s] {}
2023-01-25 21:20:17,484 INFO sqlalchemy.engine.Engine ROLLBACK
```

**Reply by cmaduike** [reply]
Thanks for the info. I should have mentioned that this is an existing table, so I don't necessarily need creation modeled here. The complication with SqlAlchemy has stemmed from defining future relations with this table in SqlAlchemy. In your example, I won't be able to create a foreign key relation because there isn't a primary key defined on the column, just on table args. I don't think sqlalchemy recognizes the table_args in such a context. This has been the one blocker I've run into.

```
class AnotherTable(Base, ModelMixin):
    __tablename__ = "another_table"

    pk = sa.Column(sa.BIGINT, primary_key=True)
    created_ts = sa.Column(sa.DateTime, nullable=True, server_default="now()")
    test_table_pk = sa.Column(sa.BIGINT, sa.ForeignKey("test_table.pk"), index=True, nullable=False)
```

I get the following error when my test harness is trying to create another_table:

`sqlalchemy.exc.ProgrammingError: (psycopg2.errors.InvalidForeignKey) there is no unique constraint matching given keys for referenced table "test_table"`

**Reply by CaselIT** [reply]
That error comes from postgresql, not sqlalchemy, so you would have to look how to make pg like that. Maybe you need to create a primary key as mentioned in https://www.postgresql.org/docs/current/ddl-partitioning.html alter index/primary key

**Reply by cmaduike** [reply]
Yes, it's complaining because in a test context where an in-memory postgres database is being created from the models, I have specified the partitioned test_table without a primary key defined on the column. When attempting to create the second table which has a column referencing test_table, the error appears. Ideally this line in your example would translate to a primary key when generating the tables in memory, but that doesn't seem to work: 

 ` __table_args__ = {"postgresql_partition_by": "RANGE (created_ts)"}`

**Reply by CaselIT** [reply]
not sure I follow. To create a foreign key you would need to make sure that the database is ok with what you are trying to do. Maybe you also need to add a pk in each partition using ddl like in your initial example `alter table test_table_2006 add primary key (pk);`

Note that if you only want to use relationships, there is no need of specifying foreign keys in the db, you can just tell sqlalchemy what condition should use for the relationship using primaryjoin.

**Reply by cmaduike** [reply]
Ok, this was a major brain fart on my part, and I now get your point. I never looked into the feasibility of referencing a partitioned table because I never had to do this in the past. It requires a constraint on the partitioning key as well, so nothing to do with SqlAlchemy. Thanks for your patience and answering my questions!

**Reply by CaselIT** [reply]
no problem! sometimes the db can be very picky in seemly obvious things
